### PR TITLE
Timer v1.0

### PR DESCRIPTION
**Pull request by Danielo814**
Here is the code we reviewed together last night. I have started trying to implement something similar to what we were talking about by separating functions but am still running into some issues.

### CODE REVIEW

File: countdown/countdown.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env node
```
Comment (usmanity) [inline countdown/countdown.js]: this is not needed in a web script, only command line node files if they're running by themselves

File: countdown/countdown.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+//     if (result <= 0) {
+//         clearInterval(myVar);
+//     } else {
+//         document.getElementById("seconds").innerHTML = result + " s";
+//     }
+
+//     }, 1000);
+   
+// }
```
Comment (usmanity) [inline countdown/countdown.js]: remove this and use git to keep your previous versions of the code

File: countdown/index.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>Timer</title>
+        <link rel="stylesheet" type="text/css" href="style.css"/>
+    </head>
+    <body>
+        <input id="input">
```
Comment (usmanity) [inline countdown/index.html]: create inputs for seconds, minutes, and hours. Also, name them appropriately.

File: countdown/countdown.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env node
+
+function getValue() {
+    var timerValue = Number(document.getElementById("input").value);
+    return timerValue;
```
Comment (usmanity) [inline countdown/countdown.js]: could this be just a one liner with a return statement?

File: countdown/countdown.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    // var myInterval = setInterval(update, 1000);
+    // startInterval();
+    // function update() {
+    //     timerValue--;
+    //     if (timerValue < 0) {
+    //         clearInterval(myInterval)
+    //     } else{
+    //        document.getElementById("seconds").innerHTML = timerValue + " s";
+    //     }
+    // }
```
Comment (usmanity) [inline countdown/countdown.js]: remove for clarity

File: countdown/countdown.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env node
+
+function getValue() {
+    var timerValue = Number(document.getElementById("input").value);
+    return timerValue;
+    // var myInterval = setInterval(update, 1000, timerValue);
+}
+
+function startInterval() {
+    var countdownInterval = setInterval(countDown, 1000, getValue());
```
Comment (usmanity) [inline countdown/countdown.js]: I'm betting you'll run into an issue with clearing the interval like this, perhaps a single function to manage interval states might be better?

**Review by usmanity — CHANGES_REQUESTED** [review]

File: countdown/countdown.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+function manageInterval() {
+    if (!window.paused) {
+        window.timerValue = getSeconds() || 1500;
+    }
+    window.countdownInterval = setInterval(function() {
+        if (timerValue < 0) {
+            clearInterval(countdownInterval);
+        } else{
```
Comment (usmanity) [inline countdown/countdown.js]: nitpick: spacing

File: countdown/countdown.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    return ((hours * 60 * 60) + (minutes * 60) + seconds); 
+}
+
+function pauseInterval() {
+    window.timerValue = timerValue;
+    window.paused = true;
+    clearInterval(countdownInterval);
+}
+
+function manageInterval() {
```
Comment (Danielo814) [inline countdown/countdown.js]: I was trying to split this function up further. But kept getting problems when trying to set the timerValue variable outside of the function that setInterval was declared in. I took note of your comment to try to manage interval states within the same function and my thinking process was that this function is doing that, but also getting the value of the timer as well.

File: countdown/countdown.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+var paused = false; 
+
+function getSeconds() {
+    hours = Number(document.getElementById("hours").value);
+    minutes = Number(document.getElementById("minutes").value);
+    seconds = Number(document.getElementById("seconds").value);
+    return ((hours * 60 * 60) + (minutes * 60) + seconds);
```
Comment (Danielo814) [inline countdown/countdown.js]: This was originally a one liner, but it looked really long and confusing, so i split it up

File: countdown/countdown.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+function getSeconds() {
+    hours = Number(document.getElementById("hours").value);
+    minutes = Number(document.getElementById("minutes").value);
+    seconds = Number(document.getElementById("seconds").value);
+    return ((hours * 60 * 60) + (minutes * 60) + seconds); 
+}
+
+function pauseInterval() {
+    window.timerValue = timerValue;
```
Comment (usmanity) [inline countdown/countdown.js]: this is technically correct but it's unnecessary. I have suggestions that might help organize this better. 

1. Either use an app name like `countdownApp` or `timerApp` and have the variables set on that as properties
2. Or declare the `window.xxx` variables outside your functions like you have declared `paused` on line 1. This way you're not fudging with the window level variables inside functions.

File: countdown/countdown.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    return document.getElementById("time").innerHTML = setTimeFormat();
+}
+
+function setValues() {
+    timerValue--;
+    return {
+        hours: Math.floor(timerValue / 60 / 60), 
+        minutes: Math.floor((timerValue / 60) % 60), 
+        seconds: Math.floor((timerValue - (Math.floor((timerValue / 60) % 60)) * 60))
+    };
```
Comment (usmanity) [inline countdown/countdown.js]: 👍 this is a good use for a return object.

File: countdown/index.html [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+        <title>Timer</title>
+        <link rel="stylesheet" type="text/css" href="style.css"/>
+    </head>
+    <body>
+        <input id="hours"> hours
+        <input id ="minutes" placeholder="25"> minutes
+        <input id="seconds" placeholder="0"> seconds
+        <button onclick="manageInterval()">Start</button>
+        <button onclick="pauseInterval()">Pause</button>
+        <p id="time"></p>
```
Comment (usmanity) [inline countdown/index.html]: for now this is fine but I'd encourage you to think about how to organize into parts that are different/similar. Like the inputs should be contained together. The buttons should live inside another container.

File: countdown/index.html [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <title>Timer</title>
```
Comment (usmanity) [inline countdown/index.html]: folder is called `countdown` title is `timer`...what is it?

File: countdown/countdown.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    }, 1000);
+}
+
+function setTimeFormat() {
+    timeUnits = setValues();
+    if (timerValue > 3600) {
+        return [timeUnits.hours + " hours  " , timeUnits.minutes + " minutes  ", timeUnits.seconds % 100 + " seconds"].join("");
+    } else {
+        return [timeUnits.minutes + " minutes  ", timeUnits.seconds % 100 + " seconds"].join("");
+    }
```
Comment (usmanity) [inline countdown/countdown.js]: Use template literals in this case, this is gnarly and will lead to issues if not already doing so: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

**Review by usmanity — CHANGES_REQUESTED** [review]
Looking a lot better but I have found a major bug which I'll let you debug before this is considered done.

**Bug:** if I keep tapping the start button, the countdown speed increases with each button click. After about 3-4 clicks, it's counting down at double speed.

**Review by usmanity — APPROVED** [review]

File: timer/countdown.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    resetButton.classList.remove("hide");
+    pauseButton.classList.remove("hide");
+    startButton.classList.add("hide");
+
+    console.log("start was clicked");
+    startCountdown();
+}
+
+function pauseHandler() {
+    var pauseButton = document.querySelector("#pauseButton");
```
Comment (jennjiang) [inline timer/countdown.js:75]: 'document.querySelector' might out of the day. Try use a jQuery selector : var pauseButton = $("#pauseButton");

**Pull request merged**